lib/libc/riscv/gen/_setjmp.S

```diff
@@ -39,7 +39,48 @@ __FBSDID("$FreeBSD$");
 
 ENTRY(_setjmp)
 #ifdef __CHERI_PURE_CAPABILITY__
-#pragma message("TODO: RISCV-PURECAP implement _setjmp")
+	/* Store the magic value and stack pointer */
+	cllc	ct0, .Lmagic
+	cld	t0, (ct0)
+	csc	ct0, (0 * 16)(ca0)
+	csc	csp, (1 * 16)(ca0)
+	cincoffset ca0, ca0, (2 * 16)
+
+	/* Store the general purpose registers and ra */
+	csc	cs0, (0 * 16)(ca0)
```

Comment:
I wonder if we should use a macro here? Something like what I did for [libunwind](https://github.com/CTSRD-CHERI/llvm-project/blob/b507d88d2aa61cec27adab60324a04b17911f5e4/libunwind/src/UnwindRegistersSave.S#L1105)?

Reply:
I'm not quite a fan of those macros for assembly. These routines are generally write-once touch-never, so the duplication of similar-but-not-quite-the-same code doesn't carry the same cost it would in code that changes.  If our libc Makefile infrastructure better supported `MACHINE_ARCH` overrides we'd probably just put the purecap versions of these in a `lib/libc/riscv64c` subdirectory where those files were preferred in `.PATH` before `lib/libc/riscv` and then these would be separate files entirely.

Note that once you fix libunwind to handle ddc you'll have a special case there anyway.  Here ddc is a slightly larger special case as it also affects the count of the offset used to adjust `a0` / `ca0`.